File: highway-pursuit-server/Server/HighwayPursuitServer.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Threading;
using System.Threading.Tasks;
using HighwayPursuitServer.Injected;

namespace HighwayPursuitServer.Server
{
    class HighwayPursuitServer
    {
        private readonly Action<string> Report;
        private readonly IHookManager _hookManager;
        private readonly EpisodeService _episodeService;
        private readonly UpdateService _updateService;
        private readonly InputService _inputService;
        private readonly Direct3D8Service _direct3D8Service;
        private readonly ScoreService _scoreService;
        private readonly CheatService _cheatService;
        private readonly Semaphore _lockUpdatePool; // Update thread waits for this
        private readonly Semaphore _lockServerPool; // Server thread waits for this

        // Game options
        const float FPS = 60.0f;
        private readonly float TICKS_PER_FRAME = Stopwatch.Frequency / FPS;
        private readonly float TICKS_PER_MS = Stopwatch.Frequency / 1000.0f;
        const long PERFORMANCE_COUNTER_FREQUENCY = 1000000;
        const int NO_REWARD_TIMEOUT = 60 * 45; // No rewards for 45 seconds result in a reset (softlock safeguard)
        const bool IS_REAL_TIME = false;
        
        // Metrics options
        const int REPORT_PERIOD = 1000;

        private long _step; // current step
        private long _lastRewardedStep; // last step where reward wans't zero
        long startTick; // used to measure performance

        public HighwayPursuitServer(Action<string> reportCallback)
        {
            // Init the update semaphore
            // Allow one game update, zero server update as the initial state
            _lockUpdatePool = new Semaphore(initialCount: 1, maximumCount: 1);
            _lockServerPool = new Semaphore(initialCount: 0, maximumCount: 1);

            // Init the report function
            Report = reportCallback;

            // Init services & hooks
            _hookManager = new HookManager(Report);
            _episodeService = new EpisodeService(_hookManager);
            _updateService = new UpdateService(_hookManager, IS_REAL_TIME, _lockServerPool, _lockUpdatePool, FPS, PERFORMANCE_COUNTER_FREQUENCY);
            _inputService = new InputService(_hookManager);
            _direct3D8Service = new Direct3D8Service(_hookManager);
            _scoreService = new ScoreService(_hookManager);
            _cheatService = new CheatService(_hookManager);

            // Activate hooks on all threads except the current thread
            _hookManager.EnableHooks();

            // Create the server thread
            CancellationTokenSource cts = new CancellationTokenSource();
            Task mainServerTask = Task.Run(() => ServerThread(cts));
            AppDomain.CurrentDomain.ProcessExit += (sender, e) =>
            {
                _hookManager.Release();
                cts.Cancel();
            };
        }

        private void ServerThread(CancellationTokenSource cts)
        {
            startTick = Environment.TickCount;

            // Here we wait for update to be called at least once
            // This ensure the game is initialized
            // We can then start a new game
            if (!cts.IsCancellationRequested)
            {
                _lockServerPool.WaitOne();
                _updateService.Step();
                _episodeService.NewGame();
                _lockUpdatePool.Release();
            }

            while (!cts.IsCancellationRequested)
            {
                if (IS_REAL_TIME)
                {
                    RealTimeLoop();
                } else
                {
                    FastLoop();
                }
            }
        }

        // Fast game update
        private void FastLoop()
        {
            HandleCurrentStep();
        }

        // Game update that is lengthened to last one frame of the specified FPS
        private void RealTimeLoop()
        {
            Stopwatch stopwatch = Stopwatch.StartNew();
            HandleCurrentStep();
            double sleepTime = (TICKS_PER_FRAME - stopwatch.ElapsedTicks) / TICKS_PER_MS;
            const int sleepToleranceMillisecond = 1;
            if (sleepTime > 0)
            {
                if (sleepTime > sleepToleranceMillisecond)
                {
                    Thread.Sleep((int)sleepTime - sleepToleranceMillisecond);
                }

                while (stopwatch.ElapsedTicks < TICKS_PER_FRAME)
                {
                    Thread.SpinWait(1);
                }
            }
            stopwatch.Stop();
        }

        // Frame code
        private void HandleCurrentStep()
        {
            // Wait for and lock game update
            _lockServerPool.WaitOne();

            // Receive action (or reset) from the gym env
            // TODO

            // Setup actions TODO: load inputs from policy
            _inputService.SetInput(new List<Input> { });

            // Get game state
            try
            {
                _direct3D8Service.Screenshot(); //TODO: how to collect/send the data?
            }
            catch (D3DERR e)
            {
                Report(e.Message);
            }

            // Get reward
            var reward = _scoreService.PullReward();
            if (reward != 0)
            {
                _lastRewardedStep = _step;
            }

            // Answer with next state/reward
            // TODO

            // Next step
            _updateService.Step();
            _step++;

            HandleSoftlock();
            HandleMetrics();

            // Release game update
            _lockUpdatePool.Release();
        }

        // Safeguard against "softlock" states that may happens
        private void HandleSoftlock()
        {
            if (_step - _lastRewardedStep > NO_REWARD_TIMEOUT)
            {
                _episodeService.NewLife();
                _lastRewardedStep = _step;
            }
        }

        // Tracks some useful metrics for debugging
        private void HandleMetrics()
        {
            // Performance metrics
            if (_step % REPORT_PERIOD == 0)
            {
                long elapsedTicks = Environment.TickCount - startTick;
                var tps = ((float)REPORT_PERIOD) / (elapsedTicks / 1000.0);
                var ratio = tps / 60.0;
                Report($"ticks/s: {tps:0} = x{ratio:0.#}");
                startTick = Environment.TickCount;
            }
        }
    }
}

File: highway-pursuit-server/highway-pursuit-server/HighwayPursuitServer.cpp
#include "pch.h"
#include "HighwayPursuitServer.hpp"

HighwayPursuitServer::HighwayPursuitServer(const Data::ServerParams& options)
    : _options(options),
    _firstEpisodeInitialized(false),
    _serverTerminated(false),
    _totalEllapsedFrames(0),
    _lastStepTermination(false, false),
    _currentInfo(Data::Info(0.0f, 0.0f, 0.0f, 0.0f)),
    _startTick(0),
    _cumulatedServerTicks(0),
    _cumulatedGameTicks(0)
{
    // num/den is the period in seconds, den/num is therefore the frequency in hertz
    auto ticks_per_s = (static_cast<float>(std::chrono::high_resolution_clock::period::den) / std::chrono::high_resolution_clock::period::num);
    TICKS_PER_FRAME = ticks_per_s / FPS;
    TICKS_PER_MS = ticks_per_s / 1000.0f;

    // Hook manager
    _hookManager = std::make_shared<HookManager>();

    // init communication manager
    _communicationManager = std::make_unique<CommunicationManager>(options);

    // init update semaphores
    _lockUpdatePool = CreateSemaphore(nullptr, 0, 1, nullptr);
    _lockServerPool = CreateSemaphore(nullptr, 0, 1, nullptr);

    // Install static hooks/services
    LARGE_INTEGER frequency;
    frequency.QuadPart = PERFORMANCE_COUNTER_FREQUENCY;
    _windowService = std::make_unique<WindowService>(_hookManager);
    _episodeService = std::make_unique<EpisodeService>(_hookManager);
    _updateService = std::make_unique<UpdateService>(_hookManager, _options.isRealTime, _lockServerPool, _lockUpdatePool, FPS, frequency);
    _scoreService = std::make_unique<ScoreService>(_hookManager);
    _cheatService = std::make_unique<CheatService>(_hookManager);

    // These need the app to be partly initialized to properly hook
    _renderingService = std::make_shared<RenderingService>(_hookManager, _options.renderParams);
    _renderingService->SetFullscreenFlag(false);
    _inputService = std::make_shared<InputService>(_hookManager);
    _hookManager->EnableHooks();
}

HighwayPursuitServer::~HighwayPursuitServer()
{
    // Try to shutdown game, release semaphore to ensure the game sees the notification
    try
    {
        _updateService->NotifyShutdown();
        ReleaseSemaphore(_lockUpdatePool, 1, nullptr);
    }
    catch (std::exception e)
    {
        HPLogger::LogException(e);
    }

    // Release all hooks/semaphores
    _hookManager->Release();
    CloseHandle(_lockUpdatePool);
    CloseHandle(_lockServerPool);
}

void HighwayPursuitServer::Run()
{
    try
    {
        // Wait for game to be initialized
        WaitGameUpdate();
        // Enable custom qpc
        _updateService->EnableCustomTime();
        SkipIntro();

        // Get server info now that the D3D device is initialized
        BufferFormat buffer = _renderingService->GetBufferFormat();
        ServerInfo serverInfo(
            buffer.height,
            buffer.width,
            buffer.channels,
            _inputService->GetInputCount()
        );

        // Setup the communication
        _communicationManager->Connect(serverInfo);

        // For performance metrics
        _startTick = GetTickCount64();

        // Main request/response loop
        while (!_serverTerminated)
        {
            auto handler = [this](InstructionCode code) { HandleInstruction(code); };
            _communicationManager->ExecuteOnInstruction(handler);
        }
    }
    // Exception handling for the server thread
    catch (HighwayPursuitException e)
    {
        HPLogger::LogException(e);
        _communicationManager->WriteException(e);
    }
    catch (std::exception e)
    {
        HPLogger::LogException(e);
        _communicationManager->WriteException(HighwayPursuitException(ErrorCode::NATIVE_ERROR));
    }
}

void HighwayPursuitServer::WaitGameUpdate()
{
    _updateService->UpdateTime();
    ReleaseSemaphore(_lockUpdatePool, 1, nullptr);
    DWORD error = WaitForSingleObject(_lockServerPool, GAME_TIMEOUT);
    if (error)
    {
        throw HighwayPursuitException(ErrorCode::GAME_TIMEOUT);
    }
}

void HighwayPursuitServer::SkipIntro()
{
     //Calling new game skips the intro
     //We then skip the initial zoom out, and wait 2 frames for the fade out the complete
     _episodeService->NewGame();
     _renderingService->ResetZoomLevel();
     WaitGameUpdate();
     WaitGameUpdate();
}

void HighwayPursuitServer::HandleInstruction(InstructionCode code)
{
    switch (code)
    {
    case InstructionCode::RESET_NEW_LIFE:
        Reset(false);
        break;
    case InstructionCode::RESET_NEW_GAME:
        Reset(true);
        break;
    case InstructionCode::STEP:
        // Return an error if step is called without reset (player dead)
        if ((!_firstEpisodeInitialized) || _lastStepTermination.IsDone())
        {
            _communicationManager->WriteNonFatalError(ErrorCode::ENVIRONMENT_NOT_RESET);
        }
        // Step at max speed or real time depending on the option
        if (_options.isRealTime)
        {
            // The frame execution inside step will be wrapped by ExecuteForOneFrame
            Step([this](std::function<void()> action) { ExecuteForOneFrame(action); });
        }
        else
        {
            Step();
        }
        break;
    case InstructionCode::CLOSE:
        // Notify end of loop
        _serverTerminated = true;
        break;
    default:
        break;
    }
}

void HighwayPursuitServer::Reset(bool startNewGame)
{
    // Force new game the first time
    if (!_firstEpisodeInitialized)
    {
        // New game is called on the first episode because it fully resets the game state
        _episodeService->NewGame();
        // Initialize the metrics
        _currentInfo = Info(0.0f, ComputeMemoryUsage(), static_cast<float>(_cumulatedServerTicks), static_cast<float>(_cumulatedGameTicks));
        _firstEpisodeInitialized = true;
    }
    // Start a new game
    else if (startNewGame)
    {
        _episodeService->NewGame();
    }
    // The game respawns the player (new life) naturally unless the last episode wasn't terminated
    else if (!_lastStepTermination.IsDone())
    {
        _episodeService->NewLife();
    }

    // Wait for one frame for the rendering buffer to update
    WaitGameUpdate();

    // Update step variables
    _lastStepTermination = Termination(false, false);

    // Return state/info
    _renderingService->Screenshot(
        [this](void* pixelData, const BufferFormat& format)
        {
            _communicationManager->WriteObservationBuffer(pixelData, format);
        });
    _communicationManager->WriteInfoBuffer(_currentInfo);
}

// TODO: this is wrong because this doesn't take into account time between instructions
void HighwayPursuitServer::ExecuteForOneFrame(std::function<void()> action)
{
    auto start = std::chrono::high_resolution_clock::now();
    action();
    auto elapsedTicks = std::chrono::duration<float>(std::chrono::high_resolution_clock::now() - start);
    double sleepTime = (TICKS_PER_FRAME - elapsedTicks.count()) / TICKS_PER_MS;
    const int sleepToleranceMillisecond = 1;

    if (sleepTime > 0)
    {
        if (sleepTime > sleepToleranceMillisecond)
        {
            std::this_thread::sleep_for(std::chrono::milliseconds(static_cast<int>(sleepTime) - sleepToleranceMillisecond));
        }

        auto duration = std::chrono::duration<float>(std::chrono::high_resolution_clock::now() - start);
        while (duration.count() * TICKS_PER_MS < TICKS_PER_FRAME)
        {
            duration = std::chrono::duration<float>(std::chrono::high_resolution_clock::now() - start);
        }
    }
    else
    {
        HPLogger::LogWarning("Step took longer than the intended number of frames.");
    }
}

void HighwayPursuitServer::Step(std::function<void(std::function<void()>)> frameWrapper)
{
    // Time spent server-side measurement
    ULONGLONG serverComputationStart = GetTickCount64();

    // Get action
    std::vector<Input> actions = _communicationManager->ReadActions();

    // Repeat action for _options.frameskip frames. Return early if episode ends.
    int cumulatedReward = 0;
    auto processFrame = [this, actions, &cumulatedReward]()
        {
            _inputService->SetInput(actions);

            // Apply action and get next state
            ULONGLONG gameComputationStart = GetTickCount64();
            WaitGameUpdate();
            _cumulatedGameTicks += GetTickCount64() - gameComputationStart; // measure time spent on running the game

            // Get reward
            int reward = _scoreService->PullReward();
            cumulatedReward += reward;

            // Next step
            _updateService->UpdateTime();
            _totalEllapsedFrames++;

            // Check termination (death)
            bool terminated = _episodeService->PullTerminated();
            _lastStepTermination = Termination(terminated, false);

            // Handle the metrics that are computed periodically
            if (_totalEllapsedFrames % PERIODIC_METRICS_FREQUENCY == 0)
            {
                ULONGLONG elapsedTicks = GetTickCount64() - _startTick;
                auto tps = PERIODIC_METRICS_FREQUENCY / (elapsedTicks / 1000.0f);

                float memorySize = ComputeMemoryUsage();
                _startTick = GetTickCount64();

                _currentInfo.memory = memorySize;
                _currentInfo.tps = tps;
            }
        };

    int skippedFrames = 0;
    while (skippedFrames < _options.frameskip && !_lastStepTermination.IsDone())
    {
        // Wrapper is here for the real time option
        if (frameWrapper == nullptr)
        {
            processFrame();
        }
        else
        {
            frameWrapper(processFrame);
        }
        skippedFrames++;
    }

    // Write return values
    _renderingService->Screenshot(
        [this](void* pixelData, const BufferFormat& format)
        {
            _communicationManager->WriteObservationBuffer(pixelData, format);
        });

    // Stop server-side timer
    _cumulatedServerTicks += GetTickCount64() - serverComputationStart;

    // Update metrics
    float serverTime = _cumulatedServerTicks / 1000.0f;
    float gameTime = _cumulatedGameTicks / 1000.0f;
    _currentInfo = Info(_currentInfo.tps, _currentInfo.memory, serverTime, gameTime);

    _communicationManager->WriteRewardBuffer(Reward(static_cast<float>(cumulatedReward)));
    _communicationManager->WriteInfoBuffer(_currentInfo);
    _communicationManager->WriteTerminationBuffer(_lastStepTermination);
}

float HighwayPursuitServer::ComputeMemoryUsage()
{
    float memorySize = 0.0f;

    HANDLE hProcess = GetCurrentProcess();
    PROCESS_MEMORY_COUNTERS pmc;

    if (GetProcessMemoryInfo(hProcess, &pmc, sizeof(pmc)))
    {
        memorySize = pmc.WorkingSetSize / (1024.0f * 1024.0f); // bytes to megabytes
    }

    return memorySize;
}